File: Matrixpy/Src/CurveFitting.py
import math

def set_data(m, x, y):
    """
    Sets the data points for curve fitting.
    Stored as an n x 2 matrix where mat[i][0] = x_i and mat[i][1] = y_i.
    """
    if len(x) != len(y) or not x:
        raise ValueError("Curve fitting data must be non-empty and x/y sizes must match.")
    m.rows = len(x)
    m.cols = 2
    m.mat = [[0.0, 0.0] for _ in range(m.rows)]
    for i in range(m.rows):
        m.mat[i][0] = float(x[i])
        m.mat[i][1] = float(y[i])

def fit_least_squares(m):
    """
    Performs Linear Least Squares fitting (y = ax + b).
    Derived from minimizing the sum of squared residuals: S = sum(y_i - (ax_i + b))^2.
    
    Normal Equations:
    1. a * sum(x^2) + b * sum(x) = sum(xy)
    2. a * sum(x)   + b * n      = sum(y)
    """
    n = m.rows
    sum_x = 0.0
    sum_y = 0.0
    sum_xx = 0.0
    sum_xy = 0.0
    
    for i in range(n):
        xi = m.mat[i][0]
        yi = m.mat[i][1]
        sum_x += xi
        sum_y += yi
        sum_xx += xi * xi
        sum_xy += xi * yi
    
    # Using Cramer's rule for a 2x2 system:
    # [ sum_xx  sum_x ] [ a ] = [ sum_xy ]
    # [ sum_x   n     ] [ b ]   [ sum_y  ]
    
    denominator = n * sum_xx - sum_x * sum_x
    if abs(denominator) < 1e-18:
        raise RuntimeError("Singular system in Least Squares (e.g., all x-values are the same).")
        
    m.a = (n * sum_xy - sum_x * sum_y) / denominator
    m.b = (sum_y * sum_xx - sum_x * sum_xy) / denominator

def evaluate_least_squares(m, x):
    """Evaluates the fitted linear model at point x."""
    return m.a * x + m.b

def rms_least_squares(m):
    """
    Calculates the Root Mean Square (RMS) Error.
    RMS = sqrt( (1/n) * sum( (y_actual - y_fitted)^2 ) )
    """
    n = m.rows
    sum_sq_err = 0.0
    for i in range(n):
        err = (m.a * m.mat[i][0] + m.b) - m.mat[i][1]
        sum_sq_err += err * err
    return math.sqrt(sum_sq_err / n)

def evaluate_range_curve(m, xMin, xMax, samples):
    """Generates (x, y) coordinates for the fitted line over a range."""
    if samples < 2:
        raise ValueError("samples must be >= 2")
    curve = []
    step = (xMax - xMin) / (samples - 1)
    for i in range(samples):
        x = xMin + i * step
        curve.append((x, m.evaluate(x)))
    return curve

def print_report_least_squares(m):
    """Prints a summary of the fitting results."""
    print("\n--- Least Squares Fitting Report ---")
    print(f"Points fitted: {m.rows}")
    print(f"Linear model:  y = {m.a:.6f}x + {m.b:.6f}")
    print(f"RMS Error:     {m.rms():.8f}")
    print("------------------------------------\n")

File: Matrix1/Src/CurveFitting.cpp
#include "CurveFitting.hpp"
#include <cmath>
#include <stdexcept>
#include <iostream>
#include <iomanip>

using namespace std;


// ================================================================
//  CurveFitting — abstract base class
// ================================================================

CurveFitting::CurveFitting() : Matrix() {}

CurveFitting::CurveFitting(const vector<long double> &xi,
                           const vector<long double> &fi)
    : Matrix()
{
    setData(xi, fi);
}

// ---------------------------------------------------------------
// setData — stores (xi, fi) into inherited Matrix as n x 2.
//   mat[i][0] = xi
//   mat[i][1] = fi   (fi = given data values)
// Same pattern as Interpolation::setData.
// ---------------------------------------------------------------
void CurveFitting::setData(const vector<long double> &xi,
                           const vector<long double> &fi)
{
    if (xi.empty() || xi.size() != fi.size())
        throw invalid_argument("CurveFitting: xi and fi must be non-empty and same size.");

    rows = static_cast<int>(xi.size());
    cols = 2;
    mat.assign(static_cast<size_t>(rows),
               vector<long double>(static_cast<size_t>(cols), 0.0L));

    for (int i = 0; i < rows; ++i)
    {
        mat[static_cast<size_t>(i)][0] = xi[static_cast<size_t>(i)]; // xi
        mat[static_cast<size_t>(i)][1] = fi[static_cast<size_t>(i)]; // fi
    }
}

int CurveFitting::pointCount() const { return rows; }

vector<long double> CurveFitting::getXi() const
{
    vector<long double> xi;
    xi.reserve(static_cast<size_t>(rows));
    for (int i = 0; i < rows; ++i)
        xi.push_back(mat[static_cast<size_t>(i)][0]);
    return xi;
}

vector<long double> CurveFitting::getFi() const
{
    vector<long double> fi;
    fi.reserve(static_cast<size_t>(rows));
    for (int i = 0; i < rows; ++i)
        fi.push_back(mat[static_cast<size_t>(i)][1]);
    return fi;
}

// Shared evaluateRange — calls the subclass evaluate() via virtual dispatch
vector<pair<long double, long double>>
CurveFitting::evaluateRange(long double xMin, long double xMax, int samples) const
{
    if (samples < 2)
        throw invalid_argument("CurveFitting: samples must be >= 2.");

    vector<pair<long double, long double>> curve;
    curve.reserve(static_cast<size_t>(samples));

    const long double step = (xMax - xMin) / static_cast<long double>(samples - 1);
    for (int i = 0; i < samples; ++i)
    {
        long double x = xMin + static_cast<long double>(i) * step;
        curve.push_back({x, evaluate(x)});
    }
    return curve;
}


// ================================================================
//  LeastSquares — concrete class
//  Linear fit: y = ax + b
//
//  From notes:
//    Minimize S = sum( (a*xi + b - fi)^2 )
//
//    dS/da = 0  =>  a*sum(xi^2) + b*sum(xi) = sum(xi*fi)
//    dS/db = 0  =>  a*sum(xi)   + b*N        = sum(fi)
//
//  Solve the 2x2 system directly (Cramer's rule):
//    det = N * sum(xi^2) - (sum(xi))^2
//    a   = ( N*sum(xi*fi) - sum(xi)*sum(fi) ) / det
//    b   = ( sum(fi)*sum(xi^2) - sum(xi)*sum(xi*fi) ) / det
// ================================================================

LeastSquares::LeastSquares() : CurveFitting(), a(0.0L), b(0.0L) {}

LeastSquares::LeastSquares(const vector<long double> &xi,
                           const vector<long double> &fi)
    : CurveFitting(xi, fi), a(0.0L), b(0.0L)
{}

// ---------------------------------------------------------------
// fit() — Normal Equations (exactly as derived in class notes)
// ---------------------------------------------------------------
void LeastSquares::fit()
{
    if (rows < 2)
        throw runtime_error("LeastSquares: need at least 2 data points.");

    const long double N = static_cast<long double>(rows);

    // Compute summations used in normal equations
    long double sum_xi    = 0.0L;  // sum(xi)
    long double sum_fi    = 0.0L;  // sum(fi)
    long double sum_xi2   = 0.0L;  // sum(xi^2)
    long double sum_xi_fi = 0.0L;  // sum(xi * fi)

    for (int i = 0; i < rows; ++i)
    {
        long double xi = mat[static_cast<size_t>(i)][0];
        long double fi = mat[static_cast<size_t>(i)][1];
        sum_xi    += xi;
        sum_fi    += fi;
        sum_xi2   += xi * xi;
        sum_xi_fi += xi * fi;
    }

    // Determinant of the 2x2 coefficient matrix
    long double det = N * sum_xi2 - sum_xi * sum_xi;

    if (fabsl(det) < 1e-15L)
        throw runtime_error("LeastSquares: singular system — cannot solve normal equations.");

    // Cramer's rule
    a = (N * sum_xi_fi  - sum_xi * sum_fi)  / det;
    b = (sum_fi * sum_xi2 - sum_xi * sum_xi_fi) / det;
}

// ---------------------------------------------------------------
// evaluate(x) — y = ax + b
// ---------------------------------------------------------------
long double LeastSquares::evaluate(long double x) const
{
    return a * x + b;
}

// ---------------------------------------------------------------
// rms() — Root Mean Square error
//   RMS = sqrt( sum( (yi - fi)^2 ) / N )
//   where yi = a*xi + b  is the fitted value
//         fi = actual data point
// ---------------------------------------------------------------
long double LeastSquares::rms() const
{
    long double sumSq = 0.0L;
    for (int i = 0; i < rows; ++i)
    {
        long double xi   = mat[static_cast<size_t>(i)][0];
        long double fi   = mat[static_cast<size_t>(i)][1];
        long double yi   = evaluate(xi);          // predicted
        long double diff = yi - fi;               // residual
        sumSq += diff * diff;
    }
    return sqrtl(sumSq / static_cast<long double>(rows));
}

// ---------------------------------------------------------------
// printReport
// ---------------------------------------------------------------
void LeastSquares::printReport() const
{
    cout << fixed << setprecision(6);
    cout << "\n--- Least Squares Fit Report ---\n";
    cout << "  Data points  N : " << rows << "\n";
    cout << "  Fitted line  : y = " << a << " * x + " << b << "\n";
    cout << "  Slope      a : " << a << "\n";
    cout << "  Intercept  b : " << b << "\n";
    cout << "  RMS Error    : " << rms() << "\n";
    cout << "--------------------------------\n";
}
